micrometer-core/src/main/java/io/micrometer/core/instrument/internal/DefaultLongTaskTimer.java

```diff
@@ -92,7 +96,9 @@ public DefaultLongTaskTimer(Id id, Clock clock, TimeUnit baseTimeUnit,
     @Override
     public Sample start() {
         SampleImpl sample = new SampleImpl();
-        activeTasks.add(sample);
+        if (!activeTasks.add(sample)) {
+            throw new IllegalStateException();
```

Comment:
We shouldn't throw exceptions when recording metrics, generally. In what case would this happen?

Reply:
It is a leftover from testing. See the implementation of SampleImpl#compareTo.
I was a bit surprised that the skip-list considers the comparator returning "0" as sufficient test for equality, and does not check Object#equals.
I am intentially leaving equality to be only reference identity (obviously two samples are never the same), but the naive comparator can not distinguish if two samples are created at exactly the same startTime. Putting hashCode in the comparison makes it "almost always" a total order, but it can theoretically still fail for hash collisions.

You can try removing the hashCode check from the comparator, and see some unit tests will fail.

A fail-proof solution would be to use a strictly increasing counter (for example an AtomicLong) instead of the time based one, but it does not look worth to increase the heap memory usage of SampleImpl.

I tend to say it is better to remove the exception that you mention here and accept that some samples can be lost (not accounted) if bothe the monotonic time does not increase and there are hash collisions.

Reply:
@shakuzen how do you suggest me to proceed?

Should I change the `DefaultLongTaskTimer.SampleImpl` class to use more heap memory and store an `AtomicInteger` to be always correct even in those very rare corner cases?

Or do you have another suggestions?

Reply:
Sorry for the late response. Now that 1.14 GA is released, I'll get back to reviewing this so we can hopefully get it merged. I'll take a look tomorrow.

Reply:
> [..] everyone would have to pay [..]

Good point, what about making only the collisions pay for storing more memory?
https://github.com/micrometer-metrics/micrometer/pull/5591/commits/d8dcd3bed4c7696c2ac64248b3d6b566dbccae00

The additional functions and indirect calls should also almost never happen, so the only disadvantage I see would be in the slightly non-obvious code